src/llvm-gcroot.cpp

```diff
@@ -0,0 +1,756 @@
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/Support/raw_ostream.h>
+#include <llvm/Transforms/Utils/BasicBlockUtils.h>
+
+#include <vector>
+#include <queue>
+
+#include "llvm-version.h"
+#include "julia.h"
+
+#ifdef LLVM37
+#define LLVM37_param(x) (x),
+#else
+#define LLVM37_param(x)
+#endif
+
+using namespace llvm;
+
+#ifndef NDEBUG
+static struct {
+    unsigned count;
+    unsigned locals;
+    unsigned temp;
+} jl_gc_frame_stats = {0};
+#endif
+
+class JuliaGCAllocator {
+public:
+    JuliaGCAllocator(CallInst *ptlsStates, Type *T_pjlvalue) :
+        F(*ptlsStates->getParent()->getParent()),
+        M(*F.getParent()),
+        T_int32(Type::getInt32Ty(F.getContext())),
+        T_int64(Type::getInt64Ty(F.getContext())),
+        V_null(Constant::getNullValue(T_pjlvalue)),
+        ptlsStates(ptlsStates),
+        gcframe(new AllocaInst(T_pjlvalue, ConstantInt::get(T_int32, 0))),
+        gcroot_func(M.getFunction("julia.gc_root_decl")),
+        gckill_func(M.getFunction("julia.gc_root_kill")),
+        gc_store_func(M.getFunction("julia.gc_store")),
+        jlcall_frame_func(M.getFunction("julia.jlcall_frame_decl"))
+    {
+/* Algorithm sketch:
+ *  Compute liveness for each basic block
+ *    liveness computed at the basic-block level for <inst, arg-offset> pairs
+ *  Propagate liveness from each basic block to its predecessors
+ *  Allocate argument slot for each jlcall frame
+ */
+#ifdef JL_DEBUG_BUILD
+        gcframe->setName("gcrootframe");
+#endif
+        gcframe->insertAfter(ptlsStates);
+        assert(gcroot_func && gckill_func && jlcall_frame_func && gc_store_func);
+    }
+
+private:
+Function &F;
+Module &M;
+Type *const T_int32;
+Type *const T_int64;
+Value *const V_null;
+CallInst *const ptlsStates;
+AllocaInst *const gcframe;
+Function *const gcroot_func;
+Function *const gckill_func;
+Function *const gc_store_func;
+Function *const jlcall_frame_func;
+
+typedef std::pair<CallInst*, unsigned> frame_register;
+class liveness {
+public:
+    typedef unsigned id;
+    enum {
+        assign = 1<<0, // an assignment to a gcroot exists in the basic-block (potentially no live-in from the predecessor basic-blocks)
+        kill   = 1<<1, // a use of a gcroot exists in the basic-block (potentially a "kill" and no live-out to the successor basic-blocks)
+        live   = 1<<2  // the gcroot is live over the entire basic-block (the assign/kill are not dominating of entry/exit)
+            // live | kill | assign == a usage and assignment exist, but it is also live on exit, the entry liveness depends on whether a store or use is encountered first
+            // live | kill == a usage exists, but the value must be live for the entire basic-block since it is not the terminal usage in the domination tree
+            // kill | assign == a usage and definition exist in domination order, so the actual lifetime is only a subset of the basic-block
+            // live | assign == impossible (this would be strange)
```

Comment:
no, the code below won't allow this case to happen. but it would me roughly that the value has an assignment, but that the previous value before the assignment needs to be live, even there is no usage of that previous value.